smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,440 @@
+            self.budget_type = 'cutoff'
+        else:
+            # budget with instances
+            self.min_budget = 1 if min_budget is None else int(min_budget)
+            self.max_budget = len(self.instances) if max_budget is None else int(max_budget)
+            self.budget_type = 'instance'
+
+            # max budget cannot be greater than number of instance-seed pairs
+            if self.max_budget > len(self.instances):
+                raise ValueError('Max budget cannot be greater than the number of instance-seed pairs')
+
+        self.logger.info("Running Successive Halving with '%s' as budget" % self.budget_type.upper())
+
+        # number configurations to consider for a full successive halving iteration
+        self.max_sh_iter = np.floor(np.log(self.max_budget / self.min_budget) / np.log(eta))
+        self.initial_challengers = int(np.round(self.eta ** (self.max_sh_iter + 1)))
+
+        # for adaptive capping
+        if self.budget_type == 'instance' and self.instance_order != 'budget_random' and self.run_obj_time:
+            self.adaptive_capping = True
+        else:
+            self.adaptive_capping = False
+        self.adaptive_capping_slackfactor = adaptive_capping_slackfactor
+
+    def intensify(self, challengers: typing.List[Configuration],
+                  incumbent: typing.Optional[Configuration],
+                  run_history: RunHistory,
+                  aggregate_func: typing.Callable,
+                  time_bound: float = float(MAXINT),
+                  log_traj: bool = True):
+        """
+        Running intensification via successive halving to determine the incumbent configuration.
+        *Side effect:* adds runs to run_history
+
+        Implementation of successive halving by Jamieson & Talwalkar (2016)
+
+        Parameters
+        ----------
+        challengers : typing.List[Configuration]
+            promising configurations
+        incumbent : Configuration
+            best configuration so far
+        run_history : RunHistory
+            stores all runs we ran so far
+        aggregate_func: typing.Callable
+            aggregate error across instances
+        time_bound : float, optional (default=2 ** 31 - 1)
+            time in [sec] available to perform intensify
+        log_traj: bool
+            whether to log changes of incumbents in trajectory
+
+        Returns
+        -------
+        incumbent: Configuration()
+            current (maybe new) incumbent configuration
+        inc_perf: float
+            empirical performance of incumbent configuration
+        """
+        self.start_time = time.time()
+        self._ta_time = 0
+        self._chall_indx = 0
+        self._num_run = 0
+
+        if time_bound < self._min_time:
+            raise ValueError("time_bound must be >= %f" % self._min_time)
+
+        if isinstance(challengers, ChallengerList):
+            challengers = challengers.challengers
```

Comment:
It was intentional because it will be easier to parallelize in the future if we filter challengers before racing them. But yes, this approach does not interleave random configurations. Will fix it. :)

Reply:
This is better, thanks, but this will introduce unnecessary overhead (sampling random configurations costs time). Please only retrieve as many challengers as you actually require below.